config/schemas/config.yaml

```diff
@@ -0,0 +1,2854 @@
+                              default: harbor
+                              type: string
+                            type: array
+                        type: object
+                      type: array
+                  type: object
+                topologyKey:
+                  default: kubernetes.io/hostname
+                  type: string
+              type: object
+            type: array
+        type: object
+    type: object
+  extraArgs:
+    items:
+      type: string
+    type: array
+  iplist:
+    items:
+      type: string
+      oneOf:
+        - format: ipv4
+        - format: ipv6
+        - format: netmask
+    type: array
+  cpumem:
+    title: CPU and Memory specifications
+    properties:
+      cpu:
+        default: 100m
+        pattern: ^[1-9][0-9]*m?$
+        type:
+          - string
+          - integer
+      memory:
+        default: 128Mi
+        pattern: ^[0-9]+([KMG]i)?$
+        type:
+          - string
+          - integer
+    type: object
+  listen:
+    properties:
+      ips:
+        $ref: '#/$defs/iplist'
+      ports:
+        $ref: '#/$defs/portlist'
+    type: object
+  nodeSelector:
+    default:
+      kubernetes.io/os: linux
+    properties:
+      kubernetes.io/os:
+        default: linux
+        type: string
+    type: object
+  portlist:
+    items:
+      maximum: 65535
+      minimum: 1
+      type: integer
+    type: array
+  resources:
+    properties:
+      limits:
+        $ref: '#/$defs/cpumem'
+      requests:
+        $ref: '#/$defs/cpumem'
+    type: object
+  tolerations:
+    items:
+      properties:
+        effect:
+          default: NoSchedule
+          type: string
+        key:
+          default: node-role.kubernetes.io/master
+          type: string
+        operator:
+          default: Exists
+          type: string
+      type: object
+    type: array
+  topologySpreadConstraints:
+    items:
+      properties:
+        labelSelector:
+          properties:
+            matchLabels:
+              type: object
+              additionalProperties:
+                type: string
+              default:
+                app.kubernetes.io/name: prometheus
+          type: object
+        maxSkew:
+          default: 1
+          type: number
+        topologyKey:
+          default: kubernetes.io/hostname
+          type: string
+        whenUnsatisfiable:
+          default: DoNotSchedule
+          type: string
+      type: object
+    type: array
+  commonResource:
+    $comment: |
+      Uncertain whether it's possible to combine $ref and additional
+      properties, short of using 'allOf'
+    type: object
+    properties:
+      affinity:
+        $ref: '#/$defs/affinity'
+      nodeSelector:
+        $ref: '#/$defs/nodeSelector'
+      resources:
+        $ref: '#/$defs/resources'
+      tolerations:
+        $ref: '#/$defs/tolerations'
+  clusterid:
+    type: string
+    enum:
+      - sc
+      - wc
+additionalProperties: false
```

Comment:
Does this mean in general that we aren't allowed to add additional fields?

Reply:
This top level one mainly serves to point out when there are top-level things that are not yet covered by the schema, since by default anything not in `properties` is allowed. Not sure if there's a way to make this the default for all `type: object` parts, that could be handy to detect typos and not yet schema'd parts of configuration.

Since `additionalProperties` is a schema (the simplest schema is a boolean), with more work on making most services have similar shape that shape could be defined as a schema here.